000X-B-C-I-R-S-error-handling-codes.md

```diff
@@ -0,0 +1,307 @@
+# Dapr Error Handling/Codes 
+
+* Author(s): Roberto J. Rojas
+* State: Draft
+* Updated: 5/11/2023
+
+## Overview
+
+Across Dapr errors are surfaced for different conditions, without consistent messages, details of the error, standard formats, and no clear indication of what/where the error initiated. 
+
+This makes troubleshooting and debugging quite difficult and requires a deep understanding of the parts of Dapr and how those parts interact with each other.
+
+To help with the issues raised above, it would be ideal if a solution could provide:
+- Greater details of errors that occured.
+- Error details in a structured format.
+- Consistency in the error details.
+- An indication where within the Dapr execution (Init, Runtime, Components, SDKs, etc...) the error occurred.
+
+## Background
+
+## Related Items
+
+### Related proposals 
+
+
+### Related issues 
+
+https://github.com/dapr/dapr/issues/6068
+
+## Expectations and alternatives
+
+## Implementation Details
+
+# Design
+
+## Solution
+Utilize and follow the [gRPC Richer Error Model](https://grpc.io/docs/guides/error/#richer-error-model) and [Google API Errors Model in the Design Guide](https://cloud.google.com/apis/design/errors#error_model)
+
+### Error Code Standard
+The [Google API Error Model](https://cloud.google.com/apis/design/errors#error_model) has the following Protobuf format:
+```go
+package google.rpc;
+
+// The `Status` type defines a logical error model that is suitable for
+// different programming environments, including REST APIs and RPC APIs.
+message Status {
+  // A simple error code that can be easily handled by the client. The
+  // actual error code is defined by `google.rpc.Code`.
+  int32 code = 1;
+
+  // A developer-facing human-readable error message in English. It should
+  // both explain the error and offer an actionable resolution to it.
+  string message = 2;
+
+  // Additional error information that the client code can use to handle
+  // the error, such as retry info or a help link.
+  repeated google.protobuf.Any details = 3;
+}
+```
+
+Here is one of the possible details that can be added to the above
+error structure. This is defined in the [error_details.proto Protobuf](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto) 
+
+```go
+message ErrorInfo {
+  // The reason of the error. This is a constant value that identifies the
+  // proximate cause of the error. Error reasons are unique within a particular
+  // domain of errors. This should be at most 63 characters and match a
+  // regular expression of `[A-Z][A-Z0-9_]+[A-Z0-9]`, which represents
+  // UPPER_SNAKE_CASE.
+  string reason = 1;
```

Comment:
is there a particular max length that is recommended for this?

Reply:
63 characters appears to be the max length. I haven't found out why, but to be on the safe side we should stick to it.